File: software/examples/delphi/ExampleCallback.pas
program ExampleCallback;

{$ifdef MSWINDOWS}{$apptype CONSOLE}{$endif}
{$ifdef FPC}{$mode OBJFPC}{$H+}{$endif}

uses
  SysUtils, IPConnection, BrickletDualButton;

type
  TExample = class
  private
    ipcon: TIPConnection;
    db: TBrickletDualButton;
  public
    procedure StateChangedCB(sender: TBrickletDualButton;                                   
                             const buttonL: byte; const buttonR: byte; 
                             const ledL: byte; const ledR: byte);
    procedure Execute;
  end;

const
  HOST = 'localhost';
  PORT = 4223;
  UID = 'XYZ'; { Change to your UID }

var
  e: TExample;

{ Callback function for state changed callback }
procedure TExample.StateChangedCB(sender: TBrickletDualButton; 
                                  const buttonL: byte; const buttonR: byte; 
                                  const ledL: byte; const ledR: byte);
begin
  if (buttonL = BRICKLET_DUAL_BUTTON_BUTTON_STATE_PRESSED) then begin
    WriteLn('Left button pressed');
  end
  else begin
    WriteLn('Left button released');
  end;

  if (buttonR = BRICKLET_DUAL_BUTTON_BUTTON_STATE_PRESSED) then begin
    WriteLn('Right button pressed');
  end
  else begin
    WriteLn('Right button released');
  end;

  WriteLn('');
end;

procedure TExample.Execute;
begin
  { Create IP connection }
  ipcon := TIPConnection.Create;

  { Create device object }
  db := TBrickletDualButton.Create(UID, ipcon);

  { Connect to brickd }
  ipcon.Connect(HOST, PORT);
  { Don't use device before ipcon is connected }

  { Register state changed callback to procedure StateChangedCB }
  db.OnStateChanged := {$ifdef FPC}@{$endif}StateChangedCB;

  WriteLn('Press key to exit');
  ReadLn;
  ipcon.Destroy; { Calls ipcon.Disconnect internally }
end;

begin
  e := TExample.Create;
  e.Execute;
  e.Destroy;
end.

File: software/examples/vbnet/ExampleCallback.vb
Imports System
Imports Tinkerforge

Module ExampleCallback
    Const HOST As String = "localhost"
    Const PORT As Integer = 4223
    Const UID As String = "XYZ" ' Change XYZ to the UID of your Dual Button Bricklet

    ' Callback subroutine for state changed callback
    Sub StateChangedCB(ByVal sender As BrickletDualButton, ByVal buttonL As Byte, _
                       ByVal buttonR As Byte, ByVal ledL As Byte, ByVal ledR As Byte)
        If buttonL = BrickletDualButton.BUTTON_STATE_PRESSED Then
            Console.WriteLine("Left Button: Pressed")
        Else If buttonL = BrickletDualButton.BUTTON_STATE_RELEASED Then
            Console.WriteLine("Left Button: Released")
        End If

        If buttonR = BrickletDualButton.BUTTON_STATE_PRESSED Then
            Console.WriteLine("Right Button: Pressed")
        Else If buttonR = BrickletDualButton.BUTTON_STATE_RELEASED Then
            Console.WriteLine("Right Button: Released")
        End If

        Console.WriteLine("")
    End Sub

    Sub Main()
        Dim ipcon As New IPConnection() ' Create IP connection
        Dim db As New BrickletDualButton(UID, ipcon) ' Create device object

        ipcon.Connect(HOST, PORT) ' Connect to brickd
        ' Don't use device before ipcon is connected

        ' Register state changed callback to subroutine StateChangedCB
        AddHandler db.StateChangedCallback, AddressOf StateChangedCB

        Console.WriteLine("Press key to exit")
        Console.ReadLine()
        ipcon.Disconnect()
    End Sub
End Module
